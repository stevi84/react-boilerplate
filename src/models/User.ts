import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
