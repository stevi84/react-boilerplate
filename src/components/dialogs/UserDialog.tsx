import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Locale } from '../../globals/Translations';
import { User, getEmptyUser } from '../../models/User';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { deleteUser, readUsers, usersSelector } from '../../reducers/UsersReducer';
import { ModalConfirmCancel } from '../common/ModalConfirmCancel';
import { DataTable } from '../dataTable/DataTable';
import { useNavigate } from 'react-router-dom';
import { Paper } from '@mui/material';
import { formatDate } from '../../globals/Formatters';
import { MainLayout } from '../common/MainLayout';

export const UserDialog = () => {
  const users: User[] = useAppSelector(usersSelector);
  const [deleteModalState, setDeleteModalState] = useState({ visible: false, entityId: -1 });
  const { t, i18n } = useTranslation();
  const lang: Locale = i18n.language as Locale;
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (users.length === 0) {
      dispatch(readUsers());
    }
  }, [dispatch, users]);

  return (
    <MainLayout>
      <Paper sx={{ margin: 1 }}>
        <DataTable<User>
          id={'user-table'}
          columns={[
            {
              field: 'name',
              headerName: t('name'),
              cellDataType: 'text',
            },
            {
              field: 'dateOfBirth',
              headerName: t('dateOfBirth'),
              cellDataType: 'text',
              valueFormatter: (params) => formatDate(params.value, lang),
            },
            {
              field: 'size',
              headerName: t('size'),
              cellDataType: 'number',
            },
            {
              field: 'weight',
              headerName: t('weight'),
              cellDataType: 'number',
            },
            {
              field: 'email',
              headerName: t('email'),
              cellDataType: 'text',
            },
            {
              field: 'phone',
              headerName: t('phone'),
              cellDataType: 'text',
            },
          ]}
          rowsData={users}
          manager={{
            create: () => navigate('/users/new'),
            read: () => dispatch(readUsers()),
            update: (entity: User) => navigate(`/users/${entity.id}`),
            delete: (entity: User) => setDeleteModalState({ visible: true, entityId: entity.id }),
            getEmpty: getEmptyUser,
          }}
        />
        <ModalConfirmCancel
          isOpen={deleteModalState.visible}
          headerText={t('delete')}
          bodyText={t('dialog_delete_text')}
          confirmButtonText={t('delete')}
          cancelButtonText={t('cancel')}
          confirmAction={() => {
            dispatch(deleteUser(deleteModalState.entityId));
            setDeleteModalState({ visible: false, entityId: -1 });
          }}
          cancelAction={() => setDeleteModalState({ visible: false, entityId: -1 })}
        />
      </Paper>
    </MainLayout>
  );
};
