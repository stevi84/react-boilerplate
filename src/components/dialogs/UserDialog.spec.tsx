import { describe, it, expect, vi, beforeEach } from 'vitest';
import { ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
import { render } from '@testing-library/react';
import { user1 } from '../../../test/data/User';
import { UserDialog } from './UserDialog';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

vi.mock('react-router-dom', () => ({ useNavigate: () => vi.fn() }));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

const { useAppSelectorMock, dispatchMock } = vi.hoisted(() => ({
  useAppSelectorMock: vi.fn(),
  dispatchMock: vi.fn(),
}));
vi.mock('../../reducers/Store', () => ({
  useAppSelector: useAppSelectorMock,
  useAppDispatch: () => dispatchMock,
}));

// ist nÃ¶tig bei Verwendung der generierten Api
const { readUsersMock } = vi.hoisted(() => ({ readUsersMock: vi.fn() }));
vi.mock('../../reducers/UsersReducer', () => ({
  deleteUser: vi.fn(),
  readUsers: readUsersMock,
  usersSelector: vi.fn(),
}));

describe('UserDialog', () => {
  beforeEach(() => {
    readUsersMock.mockReturnValue('readUsers');
  });

  it('should equal saved snapshot', () => {
    useAppSelectorMock.mockReturnValueOnce([user1]).mockReturnValue(false);
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<UserDialog />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should load users if not present', () => {
    useAppSelectorMock.mockReturnValueOnce([]).mockReturnValue(false);
    expect(dispatchMock.mock.calls.length).toEqual(0);
    render(<UserDialog />);
    expect(dispatchMock.mock.calls.length).toEqual(1);
    expect(dispatchMock.mock.calls[0][0]).toEqual('readUsers');
  });
});
