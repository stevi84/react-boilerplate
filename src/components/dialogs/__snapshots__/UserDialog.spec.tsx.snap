// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserDialog > should equal saved snapshot 1`] = `
<MainLayout>
  <ForwardRef(Paper)
    sx={
      {
        "margin": 1,
      }
    }
  >
    <DataTable
      columns={
        [
          {
            "cellDataType": "text",
            "field": "name",
            "headerName": "name",
          },
          {
            "cellDataType": "text",
            "field": "dateOfBirth",
            "headerName": "dateOfBirth",
            "valueFormatter": [Function],
          },
          {
            "cellDataType": "number",
            "field": "size",
            "headerName": "size",
          },
          {
            "cellDataType": "number",
            "field": "weight",
            "headerName": "weight",
          },
          {
            "cellDataType": "text",
            "field": "email",
            "headerName": "email",
          },
          {
            "cellDataType": "text",
            "field": "phone",
            "headerName": "phone",
          },
        ]
      }
      id="user-table"
      manager={
        {
          "create": [Function],
          "delete": [Function],
          "getEmpty": [Function],
          "read": [Function],
          "update": [Function],
        }
      }
      rowsData={
        [
          {
            "dateOfBirth": "1945-09-04T00:00:00+03:00",
            "email": "KerstinBarth@dayrep.com",
            "id": 1,
            "name": "Kerstin Barth",
            "phone": "+496563941489",
            "size": -165,
            "tsCreate": "2020-12-03T15:02:28+02:00",
            "tsUpdate": "2020-12-03T15:02:28+02:00",
            "weight": -65.56,
          },
        ]
      }
    />
    <ModalConfirmCancel
      bodyText="dialog_delete_text"
      cancelAction={[Function]}
      cancelButtonText="cancel"
      confirmAction={[Function]}
      confirmButtonText="delete"
      headerText="delete"
      isOpen={false}
    />
  </ForwardRef(Paper)>
</MainLayout>
`;
